services/authService.go

```diff
@@ -66,6 +67,27 @@ func Protect(gocloakClient *gocloak.GoCloak, authRealm string, authClientId stri
 			return
 		}
 
+		userInfo, err := gocloakClient.GetUserInfo(context.Background(), accessToken, authRealm)
+		if err != nil {
+			log.Warn().
+				Int("status", 401).
+				Str("method", r.Method).
+				Str("path", r.RequestURI).
+				Str("ip", r.RemoteAddr).
+				Str("client-id", authClientId).
+				Str("query", query).
+				Msg("Unauthorized")
+
+			w.WriteHeader(401)
+			json.NewEncoder(w).Encode(errors.BadRequestError(err.Error()))
+			return
+		}
+		username := *userInfo.PreferredUsername
+		var userClientId string = ""
+		if strings.Contains(username, "service-account-") {
+			userClientId = strings.ReplaceAll(username, "service-account-", "")
```

Comment:
Are we sure we want to remove the `service-account-` substring here? What if a service account has the same name as an actual user account; in that case, how would we tell the difference?

Reply:
@larsks I haven't found a more clever way yet, given the access token, to identify if the request is from a Keycloak Client service account, or from a user. So the only way I know is that Keycloak will always create a username that starts with "service-account-" when `serviceAccountsEnabled: true` is configured in the client. So I'll need to do some additional testing for this when we have a need for a user facing application that queries metrics with fine-grained resource permissions.